/**
 * Resolver methods for resolving remote dependencies.
 *
 * to do: only allow one concurrent resolve for every dependency `name`, but not reference.
 * would help with the semverisoning,
 * but not really
 */

(function(
  // Reliable reference to the global object (i.e. window in browsers).
  global,

  // Dummy constructor that we use as the .constructor property for
  // functions that return Generator objects.
  GeneratorFunction,

  // Undefined value, more compressible than void 0.
  undefined
) {
  var hasOwn = Object.prototype.hasOwnProperty;

  if (global.wrapGenerator) {
    return;
  }

  function wrapGenerator(innerFn, self, tryList) {
    return new Generator(innerFn, self || null, tryList || []);
  }

  global.wrapGenerator = wrapGenerator;
  if (typeof exports !== "undefined") {
    exports.wrapGenerator = wrapGenerator;
  }

  var GenStateSuspendedStart = "suspendedStart";
  var GenStateSuspendedYield = "suspendedYield";
  var GenStateExecuting = "executing";
  var GenStateCompleted = "completed";

  // Returning this object from the innerFn has the same effect as
  // breaking out of the dispatch switch statement.
  var ContinueSentinel = {};

  wrapGenerator.mark = function(genFun) {
    genFun.constructor = GeneratorFunction;
    return genFun;
  };

  // Ensure isGeneratorFunction works when Function#name not supported.
  if (GeneratorFunction.name !== "GeneratorFunction") {
    GeneratorFunction.name = "GeneratorFunction";
  }

  wrapGenerator.isGeneratorFunction = function(genFun) {
    var ctor = genFun && genFun.constructor;
    return ctor ? GeneratorFunction.name === ctor.name : false;
  };

  function Generator(innerFn, self, tryList) {
    var generator = this;
    var context = new Context(tryList);
    var state = GenStateSuspendedStart;

    function invoke(method, arg) {
      if (state === GenStateExecuting) {
        throw new Error("Generator is already running");
      }

      if (state === GenStateCompleted) {
        throw new Error("Generator has already finished");
      }

      while (true) {
        var delegate = context.delegate;
        if (delegate) {
          try {
            var info = delegate.generator[method](arg);

            // Delegate generator ran and handled its own exceptions so
            // regardless of what the method was, we continue as if it is
            // "next" with an undefined arg.
            method = "next";
            arg = undefined;

          } catch (uncaught) {
            context.delegate = null;

            // Like returning generator.throw(uncaught), but without the
            // overhead of an extra function call.
            method = "throw";
            arg = uncaught;

            continue;
          }

          if (info.done) {
            context[delegate.resultName] = info.value;
            context.next = delegate.nextLoc;
          } else {
            state = GenStateSuspendedYield;
            return info;
          }

          context.delegate = null;
        }

        if (method === "next") {
          if (state === GenStateSuspendedStart &&
              typeof arg !== "undefined") {
            // https://people.mozilla.org/~jorendorff/es6-draft.html#sec-generatorresume
            throw new TypeError(
              "attempt to send " + JSON.stringify(arg) + " to newborn generator"
            );
          }

          if (state === GenStateSuspendedYield) {
            context.sent = arg;
          } else {
            delete context.sent;
          }

        } else if (method === "throw") {
          if (state === GenStateSuspendedStart) {
            state = GenStateCompleted;
            throw arg;
          }

          if (context.dispatchException(arg)) {
            // If the dispatched exception was caught by a catch block,
            // then let that catch block handle the exception normally.
            method = "next";
            arg = undefined;
          }
        }

        state = GenStateExecuting;

        try {
          var value = innerFn.call(self, context);

          // If an exception is thrown from innerFn, we leave state ===
          // GenStateExecuting and loop back for another invocation.
          state = context.done
            ? GenStateCompleted
            : GenStateSuspendedYield;

          var info = {
            value: value,
            done: context.done
          };

          if (value === ContinueSentinel) {
            if (context.delegate && method === "next") {
              // Deliberately forget the last sent value so that we don't
              // accidentally pass it on to the delegate.
              arg = undefined;
            }
          } else {
            return info;
          }

        } catch (thrown) {
          state = GenStateCompleted;

          if (method === "next") {
            context.dispatchException(thrown);
          } else {
            arg = thrown;
          }
        }
      }
    }

    generator.next = invoke.bind(generator, "next");
    generator.throw = invoke.bind(generator, "throw");
  }

  Generator.prototype.toString = function() {
    return "[object Generator]";
  };

  function pushTryEntry(triple) {
    var entry = { tryLoc: triple[0] };

    if (1 in triple) {
      entry.catchLoc = triple[1];
    }

    if (2 in triple) {
      entry.finallyLoc = triple[2];
    }

    this.tryEntries.push(entry);
  }

  function resetTryEntry(entry, i) {
    var record = entry.completion || {};
    record.type = i === 0 ? "normal" : "return";
    delete record.arg;
    entry.completion = record;
  }

  function Context(tryList) {
    // The root entry object (effectively a try statement without a catch
    // or a finally block) gives us a place to store values thrown from
    // locations where there is no enclosing try statement.
    this.tryEntries = [{ tryLoc: "root" }];
    tryList.forEach(pushTryEntry, this);
    this.reset();
  }

  Context.prototype = {
    constructor: Context,

    reset: function() {
      this.prev = 0;
      this.next = 0;
      this.sent = undefined;
      this.done = false;
      this.delegate = null;

      this.tryEntries.forEach(resetTryEntry);

      // Pre-initialize at least 20 temporary variables to enable hidden
      // class optimizations for simple generators.
      for (var tempIndex = 0, tempName;
           hasOwn.call(this, tempName = "t" + tempIndex) || tempIndex < 20;
           ++tempIndex) {
        this[tempName] = null;
      }
    },

    stop: function() {
      this.done = true;

      var rootEntry = this.tryEntries[0];
      var rootRecord = rootEntry.completion;
      if (rootRecord.type === "throw") {
        throw rootRecord.arg;
      }

      return this.rval;
    },

    keys: function(object) {
      var keys = [];
      for (var key in object) {
        keys.push(key);
      }
      keys.reverse();

      // Rather than returning an object with a next method, we keep
      // things simple and return the next function itself.
      return function next() {
        while (keys.length) {
          var key = keys.pop();
          if (key in object) {
            next.value = key;
            next.done = false;
            return next;
          }
        }

        // To avoid creating an additional object, we just hang the .value
        // and .done properties off the next function object itself. This
        // also ensures that the minifier will not anonymize the function.
        next.done = true;
        return next;
      };
    },

    dispatchException: function(exception) {
      if (this.done) {
        throw exception;
      }

      var context = this;
      function handle(loc, caught) {
        record.type = "throw";
        record.arg = exception;
        context.next = loc;
        return !!caught;
      }

      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        var record = entry.completion;

        if (entry.tryLoc === "root") {
          // Exception thrown outside of any try block that could handle
          // it, so set the completion value of the entire function to
          // throw the exception.
          return handle("end");
        }

        if (entry.tryLoc <= this.prev) {
          var hasCatch = hasOwn.call(entry, "catchLoc");
          var hasFinally = hasOwn.call(entry, "finallyLoc");

          if (hasCatch && hasFinally) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            } else if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else if (hasCatch) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            }

          } else if (hasFinally) {
            if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else {
            throw new Error("try statement without catch or finally");
          }
        }
      }
    },

    _findFinallyEntry: function(finallyLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc <= this.prev &&
            hasOwn.call(entry, "finallyLoc") && (
              entry.finallyLoc === finallyLoc ||
              this.prev < entry.finallyLoc)) {
          return entry;
        }
      }
    },

    abrupt: function(type, arg) {
      var entry = this._findFinallyEntry();
      var record = entry ? entry.completion : {};

      record.type = type;
      record.arg = arg;

      if (entry) {
        this.next = entry.finallyLoc;
      } else {
        this.complete(record);
      }

      return ContinueSentinel;
    },

    complete: function(record) {
      if (record.type === "throw") {
        throw record.arg;
      }

      if (record.type === "break" ||
          record.type === "continue") {
        this.next = record.arg;
      } else if (record.type === "return") {
        this.rval = record.arg;
        this.next = "end";
      }

      return ContinueSentinel;
    },

    finish: function(finallyLoc) {
      var entry = this._findFinallyEntry(finallyLoc);
      return this.complete(entry.completion);
    },

    "catch": function(tryLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc === tryLoc) {
          var record = entry.completion;
          if (record.type === "throw") {
            var thrown = record.arg;
            resetTryEntry(entry, i);
          }
          return thrown;
        }
      }

      // The context.catch method must only be called with a location
      // argument that corresponds to a known catch block.
      throw new Error("illegal catch attempt");
    },

    delegateYield: function(generator, resultName, nextLoc) {
      this.delegate = {
        generator: generator,
        resultName: resultName,
        nextLoc: nextLoc
      };

      return ContinueSentinel;
    }
  };
}).apply(this, Function("return [this, function GeneratorFunction(){}]")());

var semver = require('semver');
var co = require('co');
var join = require('path').join;
var debug = require('debug')('component-resolver:dependencies');
var Resolver = require('./resolver');

/**
 * Create a dependency branch.
 *
 * @param {String} branch
 * @param {String} repo
 * @return {Object} branch
 * @api private
 */

Resolver.prototype.branchDependency = wrapGenerator.mark(function(name, ref, parent) {
  var slug, state, i, dep, branch, remotes, remote;

  return wrapGenerator(function($ctx0) {
    while (1) switch ($ctx0.prev = $ctx0.next) {
    case 0:
      slug = name + '@' + ref;
      state = this.state[slug];

      if (!(state === 'resolving')) {
        $ctx0.next = 8;
        break;
      }

      $ctx0.next = 5;
      return this.await(slug);
    case 5:
      return $ctx0.abrupt("return", $ctx0.sent);
    case 8:
      if (!(state === 'resolved')) {
        $ctx0.next = 18;
        break;
      }

      i = 0;
    case 10:
      if (!(i < this.dependencies.length)) {
        $ctx0.next = 17;
        break;
      }

      dep = this.dependencies[i];

      if (!(dep.name === name && dep.ref === ref)) {
        $ctx0.next = 14;
        break;
      }

      return $ctx0.abrupt("return", dep);
    case 14:
      i++;
      $ctx0.next = 10;
      break;
    case 17:
      throw new Error('wtf');
    case 18:
      debug('resolving dependency %s@%s', name, ref);
      this.status[slug] = true;

      // mark as resolving
      this.state[slug] = 'resolving';

      branch = {
        type: 'dependency',
        name: name,
        ref: ref,
        canonical: slug.replace('/', '~'),
        parent: parent,
        version: semver.valid(ref),
        dependencies: {},
        locals: {}, // always empty, but added for consistency
        dependents: []
      };

      remotes = this.resolveRemotes(branch);
      debug('searching %s for %s@%s', JSON.stringify(remotes), name, ref);
      return $ctx0.delegateYield(this.remote.resolve(remotes, name, ref), "t0", 25);
    case 25:
      remote = $ctx0.t0;

      if (!(!remote && semver.valid(ref) && ref[0] !== 'v')) {
        $ctx0.next = 31;
        break;
      }

      ref = branch.ref = 'v' + ref;
      return $ctx0.delegateYield(this.remote.resolve(remotes, name, ref), "t1", 29);
    case 29:
      remote = $ctx0.t1;
      if (!remote) ref = branch.ref = ref.slice(1);
    case 31:
      if (!!remote) {
        $ctx0.next = 33;
        break;
      }

      throw new Error('no remote found for dependency "' + name + '@' + ref + '". Visit http://component.github.io/troubleshooting for help.');
    case 33:
      return $ctx0.delegateYield(remote.json(name, ref), "t2", 34);
    case 34:
      branch.node = $ctx0.t2;
      debug('found %s@%s from remote "%s"', name, ref, remote.name);

      if (!!branch.node) {
        $ctx0.next = 38;
        break;
      }

      throw new Error('trouble finding dependency "' + slug + '".');
    case 38:
      branch.path = join(this.out, name, ref);
      branch.filename = join(branch.path, 'component.json');
      this.dependencies.push(branch);
      this.resolve(branch);
      this.state[slug] = 'resolved';
      this.emit(name, branch);
      this.emit(slug, branch);
      if (this.install) this.download(branch);
      delete this.status[slug];
      return $ctx0.abrupt("return", branch);
    case 48:
    case "end":
      return $ctx0.stop();
    }
  }, this);
});

/**
 * Download a component.
 *
 * @param {Object} branch
 * @api private
 */

Resolver.prototype.download = function (branch) {
  var task = this._download(branch.resolvedRemotes, branch.name, branch.ref);
  this.channel.installs.push(co(task));
}

/**
 * Resolve an object of dependencies.
 *
 * @param {Object} branch
 * @param {Object} dependencies
 * @api private
 */

Resolver.prototype.resolveDependencies = function (branch, deps) {
  if (!deps) return
  var ch = this.channel.dependencies
  Object.keys(deps).forEach(function (name) {
    ch.push(co(this.resolveDependency(branch, name.toLowerCase(), deps[name])))
  }, this)
}

/**
 * Resolve a branch's dependency.
 *
 * to do: remotes
 *
 * @param {Object} branch
 * @param {String} repo
 * @param {String} ref
 * @api {private}
 */

Resolver.prototype.resolveDependency = wrapGenerator.mark(function(branch, name, ref) {
  var child, version;

  return wrapGenerator(function($ctx1) {
    while (1) switch ($ctx1.prev = $ctx1.next) {
    case 0:
      child = branch.dependencies[name];

      if (!(typeof child === 'object')) {
        $ctx1.next = 3;
        break;
      }

      return $ctx1.abrupt("return");
    case 3:
      if (!(typeof child === 'string' && !this.resolvingSemver)) {
        $ctx1.next = 5;
        break;
      }

      return $ctx1.abrupt("return");
    case 5:
      if (!(semver.validRange(ref) && !semver.valid(ref))) {
        $ctx1.next = 15;
        break;
      }

      if (!this.resolvingSemver) {
        $ctx1.next = 9;
        break;
      }

      return $ctx1.delegateYield(this.resolveSemver(branch, name, ref), "t3", 8);
    case 8:
      return $ctx1.abrupt("return", $ctx1.t3);
    case 9:
      version = this.resolveSemverLocally(branch, name, ref);

      if (!version) {
        $ctx1.next = 14;
        break;
      }

      // we'll resolve this later
      // to do: try every time a repo by `name` is resolved
      // until one finally matches
      ref = version;

      $ctx1.next = 15;
      break;
    case 14:
      return $ctx1.abrupt("return", branch.dependencies[name] = ref);
    case 15:
      return $ctx1.delegateYield(this.branchDependency(name, ref, branch), "t4", 16);
    case 16:
      child = $ctx1.t4;
      branch.dependencies[name] = child;
      if (!~child.dependents.indexOf(branch)) child.dependents.push(branch);
      debug('resolved dependencies of "%s"', branch.name);
    case 20:
    case "end":
      return $ctx1.stop();
    }
  }, this);
});

/**
 * Get the currently resolved references of a dependency.
 *
 * @param {String} name
 * @return {Array} references
 * @api private
 */

Resolver.prototype.getReferences = function (name) {
  return this.dependencies.filter(function (dep) {
    return dep.name === name
  }).map(getRef)
}

function getRef(x) {
  return x.ref
}